File: bindings/js/tests/fixture.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { ProjectionSession, ProjectionError } from "../src/index.js";
import type { EmittedEvent } from "../src/index.js";

interface FixtureError {
	code: string;
	description?: string;
}

import type { SessionOptions } from "../src/types.js";

type DiagnosticSeverityName = "error" | "warning" | "information" | "hint";

interface FixtureDiagnostic {
	code: string;
	severity?: DiagnosticSeverityName;
}

interface Fixture {
	name: string;
	source: string;
	options?: SessionOptions;
	setState?: { partition: string | null; state: string };
	events?: Record<string, unknown>[];
	expect: {
		sources?: Record<string, unknown>;
		state?: unknown;
		states?: Record<string, unknown>;
		sharedState?: unknown;
		result?: unknown;
		emitted?: { streamId: string; eventType: string; data: string }[];
		logs?: string[];
		error?: FixtureError;
		getResult?: boolean;
		diagnostics?: FixtureDiagnostic[];
	};
}

const SEVERITY_BY_NAME: Record<DiagnosticSeverityName, number> = {
	error: 1,
	warning: 2,
	information: 3,
	hint: 4,
};

function loadFixtures(filename: string): Fixture[] {
	const path = join(__dirname, "..", "..", "..", "tools", "fixtures", filename);
	return JSON.parse(readFileSync(path, "utf8"));
}

function runFixtures(filename: string) {
	const fixtures = loadFixtures(filename);
	for (const f of fixtures) {
		it(f.name, () => runFixture(f));
	}
}

function assertError(error: ProjectionError, expected: FixtureError) {
	expect(error.code).toBe(expected.code);
	if (expected.description) {
		expect(error.description).toContain(expected.description);
	}
}

function runFixture(f: Fixture) {
	if (!f.options) {
		throw new Error(
			`Fixture "${f.name}" missing required options (engineVersion)`,
		);
	}

	// Creation error (no events, no getResult)
	if (f.expect.error && !f.events?.length && !f.expect.getResult) {
		try {
			new ProjectionSession(f.source, f.options);
			expect.fail("Expected error but session created successfully");
		} catch (err) {
			expect(err).toBeInstanceOf(ProjectionError);
			assertError(err as ProjectionError, f.expect.error);
		}
		return;
	}

	const session = new ProjectionSession(f.source, f.options);

	try {
		// Check sources
		if (f.expect.sources) {
			const sources = session.getSources();
			for (const [key, expected] of Object.entries(f.expect.sources)) {
				expect((sources as unknown as Record<string, unknown>)[key]).toEqual(
					expected,
				);
			}
		}

		// Check diagnostics: strict count match AND each expected entry
		// must consume a distinct actual entry (one-to-one). Without the
		// consume step, an expected [A, A] would match an actual [A, B]
		// because both expected entries find the first A.
		if (f.expect.diagnostics) {
			const remaining = [...(session.getSources().diagnostics ?? [])];
			expect(remaining.length).toBe(f.expect.diagnostics.length);
			for (const exp of f.expect.diagnostics) {
				const expectedSeverity = exp.severity
					? SEVERITY_BY_NAME[exp.severity]
					: undefined;
				const matchIndex = remaining.findIndex(
					(d) =>
						d.code === exp.code &&
						(expectedSeverity === undefined || d.severity === expectedSeverity),
				);
				if (matchIndex < 0) {
					expect.fail(
						`expected diagnostic with code ${exp.code}` +
							(exp.severity ? ` / severity ${exp.severity}` : "") +
							`; remaining: [${remaining.map((d) => d.code).join(", ")}]`,
					);
				}
				remaining.splice(matchIndex, 1);
			}
		}

		// Set state
		if (f.setState) {
			session.setState(f.setState.partition, f.setState.state);
		}

		// Feed events
		if (f.events?.length) {
			let lastFeedError: ProjectionError | null = null;
			let lastEmitted: EmittedEvent[] = [];
			let lastLogs: string[] = [];

			session.onEmit((event) => {
				lastEmitted.push(event);
			});
			session.onLog((message) => {
				lastLogs.push(message);
			});

			for (const event of f.events) {
				lastEmitted = [];
				lastLogs = [];

				try {
					// eslint-disable-next-line @typescript-eslint/no-explicit-any
					session.feed(event as any);
				} catch (err) {
					if (err instanceof ProjectionError) {
						lastFeedError = err;
					} else {
						throw err;
					}
				}
			}

			// Check feed error (not getResult)
			if (f.expect.error && !f.expect.getResult) {
				expect(lastFeedError).not.toBeNull();
				if (!lastFeedError) return;
				assertError(lastFeedError, f.expect.error);
				return;
			}

			// Check emitted
			if (f.expect.emitted !== undefined) {
				expect(lastEmitted).toHaveLength(f.expect.emitted.length);
				for (const [i, expected] of f.expect.emitted.entries()) {
					const actual = lastEmitted[i];
					expect(actual).toBeDefined();
					if (!actual) continue;
					expect(actual.streamId).toBe(expected.streamId);
					expect(actual.eventType).toBe(expected.eventType);
					if (expected.data) {
						expect(actual.data).toBe(expected.data);
					}
				}
			}

			// Check logs
			if (f.expect.logs !== undefined) {
				expect(lastLogs).toEqual(f.expect.logs);
			}
		}

		// Check getResult error
		if (f.expect.getResult && f.expect.error) {
			try {
				session.getResult();
				expect.fail("Expected error but getResult succeeded");
			} catch (err) {
				expect(err).toBeInstanceOf(ProjectionError);
				assertError(err as ProjectionError, f.expect.error);
			}
			return;
		}

		// Check state
		if (f.expect.state !== undefined) {
			const state = session.getStateJson();
			expect(state).toEqual(f.expect.state);
		}

		// Check per-partition states
		if (f.expect.states) {
			for (const [partition, expected] of Object.entries(f.expect.states)) {
				if (expected === null) {
					expect(session.getState(partition)).toBeNull();
				} else {
					const state = session.getStateJson(partition);
					expect(state).toEqual(expected);
				}
			}
		}

		// Check shared state
		if (f.expect.sharedState !== undefined) {
			expect(session.getSharedStateJson()).toEqual(f.expect.sharedState);
		}

		// Check result
		if (f.expect.result !== undefined) {
			if (f.expect.result === null) {
				expect(session.getResult()).toBeNull();
			} else {
				expect(session.getResultJson()).toEqual(f.expect.result);
			}
		}
	} finally {
		session.dispose();
	}
}

describe("Fixtures: sources", () => runFixtures("sources.json"));
describe("Fixtures: state", () => runFixtures("state.json"));
describe("Fixtures: callbacks", () => runFixtures("callbacks.json"));
describe("Fixtures: errors", () => runFixtures("errors.json"));
describe("Fixtures: transforms", () => runFixtures("transforms.json"));
describe("Fixtures: deletion", () => runFixtures("deletion.json"));
describe("Fixtures: versioning", () => runFixtures("versioning.json"));

File: bindings/go/fixture_test.go
package gafferruntime

import (
	"encoding/json"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"testing"
)

type fixture struct {
	Name     string          `json:"name"`
	Source   string          `json:"source"`
	Options  json.RawMessage `json:"options,omitempty"`
	SetState *struct {
		Partition *string `json:"partition"`
		State     string  `json:"state"`
	} `json:"setState,omitempty"`
	Events []json.RawMessage `json:"events,omitempty"`
	Expect fixtureExpect     `json:"expect"`
}

type fixtureExpect struct {
	Sources     map[string]json.RawMessage `json:"sources,omitempty"`
	State       json.RawMessage            `json:"state,omitempty"`
	States      map[string]json.RawMessage `json:"states,omitempty"`
	SharedState json.RawMessage            `json:"sharedState,omitempty"`
	Result      json.RawMessage            `json:"result,omitempty"`
	Emitted     []fixtureEmitted           `json:"emitted,omitempty"`
	Logs        []string                   `json:"logs,omitempty"`
	Error       *fixtureError              `json:"error,omitempty"`
	GetResult   *bool                      `json:"getResult,omitempty"`
}

type fixtureError struct {
	Code        string `json:"code"`
	Description string `json:"description,omitempty"`
}

type fixtureEmitted struct {
	StreamID  string `json:"streamId"`
	EventType string `json:"eventType"`
	Data      string `json:"data"`
}

func loadFixtures(t *testing.T, filename string) []fixture {
	t.Helper()
	path := filepath.Join("..", "..", "tools", "fixtures", filename)
	data, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("failed to read fixtures: %v", err)
	}
	var fixtures []fixture
	if err := json.Unmarshal(data, &fixtures); err != nil {
		t.Fatalf("failed to parse fixtures: %v", err)
	}
	return fixtures
}

func runFixtures(t *testing.T, filename string) {
	t.Helper()
	fixtures := loadFixtures(t, filename)
	for _, f := range fixtures {
		t.Run(f.Name, func(t *testing.T) {
			runFixture(t, f)
		})
	}
}

func runFixture(t *testing.T, f fixture) {
	t.Helper()

	var opts *string
	if f.Options != nil {
		s := string(f.Options)
		opts = &s
	}

	// Creation error (no events, no getResult)
	if f.Expect.Error != nil && len(f.Events) == 0 && (f.Expect.GetResult == nil || !*f.Expect.GetResult) {
		_, err := NewSession(f.Source, opts)
		if err == nil {
			t.Fatal("expected error on create")
		}
		assertFixtureError(t, err, f.Expect.Error)
		return
	}

	session, err := NewSession(f.Source, opts)
	if err != nil {
		t.Fatalf("NewSession failed: %v", err)
	}
	defer session.Destroy()

	// Check sources
	if f.Expect.Sources != nil {
		sourcesJSON := session.GetSources()
		if sourcesJSON == nil {
			t.Fatal("GetSources returned nil")
		}
		assertSourcesMatch(t, *sourcesJSON, f.Expect.Sources)
	}

	// Set state
	if f.SetState != nil {
		session.SetState(f.SetState.Partition, f.SetState.State)
	}

	// Feed events
	if len(f.Events) > 0 {
		var lastEmitted []struct{ streamID, eventType, data string }
		var lastLogs []string
		var lastFeedErr error

		session.OnEmit(func(streamID, eventType, data, _ string, _, _ bool) {
			lastEmitted = append(lastEmitted, struct{ streamID, eventType, data string }{streamID, eventType, data})
		})
		session.OnLog(func(message string) {
			lastLogs = append(lastLogs, message)
		})

		for _, evRaw := range f.Events {
			lastEmitted = nil
			lastLogs = nil
			if feedErr := session.Feed(string(evRaw)); feedErr != nil {
				lastFeedErr = feedErr
			}
		}

		// Check feed error (not getResult)
		if f.Expect.Error != nil && (f.Expect.GetResult == nil || !*f.Expect.GetResult) {
			if lastFeedErr == nil {
				t.Fatal("expected feed error")
			}
			assertFixtureError(t, lastFeedErr, f.Expect.Error)
			return
		}

		// Check emitted
		if f.Expect.Emitted != nil {
			if len(f.Expect.Emitted) != len(lastEmitted) {
				t.Fatalf("expected %d emitted events, got %d", len(f.Expect.Emitted), len(lastEmitted))
			}
			for i, exp := range f.Expect.Emitted {
				act := lastEmitted[i]
				if exp.StreamID != act.streamID {
					t.Fatalf("emitted[%d] streamId: expected %q, got %q", i, exp.StreamID, act.streamID)
				}
				if exp.EventType != act.eventType {
					t.Fatalf("emitted[%d] eventType: expected %q, got %q", i, exp.EventType, act.eventType)
				}
				if exp.Data != "" && exp.Data != act.data {
					t.Fatalf("emitted[%d] data: expected %q, got %q", i, exp.Data, act.data)
				}
			}
		}

		// Check logs
		if f.Expect.Logs != nil {
			if len(f.Expect.Logs) != len(lastLogs) {
				t.Fatalf("expected %d logs, got %d", len(f.Expect.Logs), len(lastLogs))
			}
			for i, exp := range f.Expect.Logs {
				if exp != lastLogs[i] {
					t.Fatalf("log[%d]: expected %q, got %q", i, exp, lastLogs[i])
				}
			}
		}
	}

	// Check getResult error
	if f.Expect.GetResult != nil && *f.Expect.GetResult && f.Expect.Error != nil {
		_, err := session.GetResult(nil)
		if err == nil {
			t.Fatal("expected getResult error")
		}
		assertFixtureError(t, err, f.Expect.Error)
		return
	}

	// Check state
	if f.Expect.State != nil {
		state := session.GetState(nil)
		if state == nil {
			t.Fatal("GetState returned nil")
		}
		assertJSONEqual(t, "state", string(f.Expect.State), *state)
	}

	// Check per-partition states
	for partition, expected := range f.Expect.States {
		p := partition
		state := session.GetState(&p)
		if string(expected) == "null" {
			if state != nil {
				t.Fatalf("expected nil state for partition %q, got %s", partition, *state)
			}
		} else {
			if state == nil {
				t.Fatalf("expected state for partition %q, got nil", partition)
			}
			assertJSONEqual(t, fmt.Sprintf("state[%s]", partition), string(expected), *state)
		}
	}

	// Check shared state
	if f.Expect.SharedState != nil {
		shared := session.GetSharedState()
		if shared == nil {
			t.Fatal("GetSharedState returned nil")
		}
		assertJSONEqual(t, "sharedState", string(f.Expect.SharedState), *shared)
	}

	// Check result
	if f.Expect.Result != nil {
		if string(f.Expect.Result) == "null" {
			result, err := session.GetResult(nil)
			if err != nil {
				t.Fatalf("GetResult error: %v", err)
			}
			if result != nil {
				t.Fatalf("expected nil result, got %s", *result)
			}
		} else {
			result, err := session.GetResult(nil)
			if err != nil {
				t.Fatalf("GetResult error: %v", err)
			}
			if result == nil {
				t.Fatal("GetResult returned nil")
			}
			assertJSONEqual(t, "result", string(f.Expect.Result), *result)
		}
	}
}

func assertFixtureError(t *testing.T, err error, expected *fixtureError) {
	t.Helper()
	var ge GafferError
	if !errors.As(err, &ge) {
		t.Fatalf("expected GafferError, got %T: %v", err, err)
	}
	if ge.ErrorCode() != expected.Code {
		t.Fatalf("expected code %q, got %q", expected.Code, ge.ErrorCode())
	}
	if expected.Description != "" && !strings.Contains(ge.ErrorDescription(), expected.Description) {
		t.Fatalf("expected description containing %q, got %q", expected.Description, ge.ErrorDescription())
	}
}

func assertJSONEqual(t *testing.T, label, expected, actual string) {
	t.Helper()
	var exp, act interface{}
	if err := json.Unmarshal([]byte(expected), &exp); err != nil {
		t.Fatalf("%s: invalid expected JSON: %v", label, err)
	}
	if err := json.Unmarshal([]byte(actual), &act); err != nil {
		t.Fatalf("%s: invalid actual JSON: %v", label, err)
	}
	expNorm, _ := json.Marshal(exp)
	actNorm, _ := json.Marshal(act)
	if string(expNorm) != string(actNorm) {
		t.Fatalf("%s:\n  expected: %s\n  actual:   %s", label, expNorm, actNorm)
	}
}

func assertSourcesMatch(t *testing.T, sourcesJSON string, expected map[string]json.RawMessage) {
	t.Helper()
	var sources map[string]json.RawMessage
	if err := json.Unmarshal([]byte(sourcesJSON), &sources); err != nil {
		t.Fatalf("failed to parse sources JSON: %v", err)
	}
	for key, expVal := range expected {
		actVal, ok := sources[key]
		if !ok {
			t.Fatalf("sources missing key %q", key)
		}
		expNorm, _ := json.Marshal(json.RawMessage(expVal))
		actNorm, _ := json.Marshal(json.RawMessage(actVal))
		if string(expNorm) != string(actNorm) {
			t.Fatalf("sources[%s]:\n  expected: %s\n  actual:   %s", key, expNorm, actNorm)
		}
	}
}

func TestFixtures_Sources(t *testing.T)    { runFixtures(t, "sources.json") }
func TestFixtures_State(t *testing.T)      { runFixtures(t, "state.json") }
func TestFixtures_Callbacks(t *testing.T)  { runFixtures(t, "callbacks.json") }
func TestFixtures_Errors(t *testing.T)     { runFixtures(t, "errors.json") }
func TestFixtures_Transforms(t *testing.T) { runFixtures(t, "transforms.json") }
func TestFixtures_Deletion(t *testing.T)   { runFixtures(t, "deletion.json") }
func TestFixtures_Versioning(t *testing.T) { runFixtures(t, "versioning.json") }
